### PR TITLE
Custom CheckBox

### PR DESCRIPTION
**Pull request by 84634E1A607A**
A custom CheckBox that changes icon when mouse hovers / disabled.

**Related Issues**

Issue comment: [2459](https://github.com/Revolutionary-Games/Thrive/pull/2459/#issuecomment-906079447)

**Progress Checklist**

Note: before starting this checklist the issue should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
Could you rebase this PR so that this only has the new commits? Would be much easier to review this that way...

**Comment by 84634E1A607A** [comment]
Costs some time, but completed.

File: src/gui_common/TweakedCheckBox.tscn [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[gd_scene load_steps=9 format=2]
+
+[ext_resource path="res://src/gui_common/fonts/Jura-DemiBold-Smaller.tres" type="DynamicFont" id=1]
```
Comment (Kasterisk) [inline src/gui_common/TweakedCheckBox.tscn:3]: I much prefer Jura-DemiBold-AlmostSmaller.tres if you want a smaller font as this one is too small.

File: src/gui_common/TweakedColourPicker.cs [hunk 2]
```diff
@@ -121,10 +120,6 @@ public new bool RawMode
         set
         {
             mode = value ? PickerMode.Raw : PickerMode.Rgb;
-
-            if (value == false)
-                ValidateRgbColor();
-
```
Comment (Kasterisk) [inline src/gui_common/TweakedColourPicker.cs:127]: Unintended change?

File: src/gui_common/TweakedColourPicker.cs [hunk 3]
```diff
@@ -303,23 +297,6 @@ private void OnGroupErasePreset(Color colour)
...
-        if (colour.r > 1)
-            colour.r = 1;
-        if (colour.g > 1)
-            colour.g = 1;
-        if (colour.b > 1)
-            colour.b = 1;
-
-        Color = colour;
-    }
-
```
Comment (Kasterisk) [inline src/gui_common/TweakedColourPicker.cs:322]: Unintended change?

File: src/gui_common/TweakedColourPicker.cs [hunk 4]
```diff
@@ -401,9 +378,6 @@ private void OnRawButtonToggled(bool isOn)
 
     private void OnColourChanged(Color colour)
     {
-        // Hide HtmlColourEdit color change when Raw mode is on and any color value is above 1.0
-        htmlColourStart.Visible = htmlColourEdit.Visible = !(mode == PickerMode.Raw && colour.IsRaw());
-
```
Comment (Kasterisk) [inline src/gui_common/TweakedColourPicker.cs:406]: Unintended change?

File: src/gui_common/TweakedColourPicker.cs [hunk 5]
```diff
@@ -489,8 +463,7 @@ public override void _GuiInput(InputEvent inputEvent)
         private void UpdateTooltip()
         {
             HintTooltip = string.Format(CultureInfo.CurrentCulture,
-                TranslationServer.Translate("COLOUR_PICKER_PRESET_TOOLTIP"),
-                Color.IsRaw() ? "argb(" + Color + ")" : "#" + Color.ToHtml());
```
Comment (Kasterisk) [inline src/gui_common/TweakedColourPicker.cs:493]: These are undoing the raw color handling for html color edit.

File: src/gui_common/TweakedColourPicker.tscn [hunk 6]
```diff
@@ -115,14 +110,14 @@ columns = 12
...
 custom_constants/margin_top = 5
 
 [node name="AddPresetButton" type="TextureButton" parent="PresetButtonContainer"]
 margin_top = 5.0
-margin_right = 25.0
-margin_bottom = 30.0
-rect_min_size = Vector2( 25, 25 )
+margin_right = 30.0
+margin_bottom = 35.0
+rect_min_size = Vector2( 30, 30 )
```
Comment (Kasterisk) [inline src/gui_common/TweakedColourPicker.tscn:120]: Add preset button's size is changed from previous.

File: src/gui_common/TweakedColourPicker.tscn [hunk 7]
```diff
@@ -31,48 +32,37 @@ margin_left = 10.0
 margin_top = 5.0
 margin_right = 289.0
 margin_bottom = 32.0
-rect_min_size = Vector2( 0, 27 )
```
Comment (Kasterisk) [inline src/gui_common/TweakedColourPicker.tscn:34]: This is probably also a collateral damage.

File: src/gui_common/TweakedColourPicker.tscn [hunk 8]
```diff
@@ -81,25 +71,30 @@ __meta__ = {
...
+margin_left = 167.0
+margin_right = 260.0
 margin_bottom = 27.0
 size_flags_horizontal = 3
 theme = ExtResource( 1 )
 text = "00000000"
 max_length = 8
 context_menu_enabled = false
 
+[node name="CheckBox" parent="MarginButtonsContainer/ButtonsContainer" instance=ExtResource( 6 )]
```
Comment (Kasterisk) [inline src/gui_common/TweakedColourPicker.tscn:93]: Random checkbox next to the html textbox.

File: src/microbe_stage/editor/MicrobeEditor.tscn [hunk 9]
```diff
@@ -3432,8 +3428,10 @@ __meta__ = {
 [connection signal="mouse_entered" from="MicrobeEditorGUI/TopLeft/HBoxContainer/CellEditorButton" to="MicrobeEditorGUI" method="OnMouseEnter"]
 [connection signal="mouse_exited" from="MicrobeEditorGUI/TopLeft/HBoxContainer/CellEditorButton" to="MicrobeEditorGUI" method="OnMouseExit"]
 [connection signal="pressed" from="MicrobeEditorGUI/TopLeft/HBoxContainer/CellEditorButton" to="MicrobeEditorGUI" method="SetEditorTab" binds= [ "CellEditor" ]]
-[connection signal="mouse_entered" from="MicrobeEditorGUI/CellEditor/TopRight" to="MicrobeEditorGUI" method="OnMouseEnter"]
-[connection signal="mouse_exited" from="MicrobeEditorGUI/CellEditor/TopRight" to="MicrobeEditorGUI" method="OnMouseExit"]
```
Comment (Kasterisk) [inline src/microbe_stage/editor/MicrobeEditor.tscn:3436]: Disconnected signals.

File: src/gui_common/TweakedCheckBox.cs [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        UpdateIcon();
+        base._Process(delta);
+    }
+
+    public override void _GuiInput(InputEvent @event)
+    {
+        if (@event is InputEventMouseButton { ButtonIndex: (int)ButtonList.Left } mouseEvent)
+        {
+            pressing = mouseEvent.Pressed;
+        }
```
Comment (Kasterisk) [inline src/gui_common/TweakedCheckBox.cs:35]: Could have a comment why this is.

**Review by Kasterisk — CHANGES_REQUESTED** [review]
I see this is moving away from the default highlight area on hover to a more dynamic checked icons which is likely what DonGororo originally intended for the checkboxes when making the icons.

There should be a bit more spacing between the icon and the text:
checkbox

File: src/gui_common/TweakedCheckBox.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        unpressedHover = GD.Load<Texture>("res://assets/textures/gui/bevel/checkAhover.png");
+        unpressedClick = GD.Load<Texture>("res://assets/textures/gui/bevel/checkAclick.png");
+        pressedNormal = GD.Load<Texture>("res://assets/textures/gui/bevel/checkB.png");
+        pressedHover = GD.Load<Texture>("res://assets/textures/gui/bevel/checkBhover.png");
+        pressedClick = GD.Load<Texture>("res://assets/textures/gui/bevel/checkBclick.png");
+    }
+
+    public override void _Process(float delta)
+    {
+        UpdateIcon();
```
[the remark this answers is not in the document]
Comment (84634E1A607A) [reply src/gui_common/TweakedCheckBox.cs:26]: No, after overriding _Draw(), Godot always draw the last icon I set, not the current one.

File: src/microbe_stage/editor/MicrobeEditor.tscn (hunk 9 above)
Comment (84634E1A607A) [reply src/microbe_stage/editor/MicrobeEditor.tscn:3436]: I hate merge conflicts.

**Comment by 84634E1A607A** [comment]
> There should be a bit more spacing between the icon and the text:
> checkbox

Icon is a Godot property. I don't think I can change that.

**Comment by 84634E1A607A** [comment]
image

Look better? Added 3 pixels

File: src/gui_common/TweakedCheckBox.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Kasterisk) [reply src/gui_common/TweakedCheckBox.cs]: Huh, I guess I'm wrong, the internal redraw call after setting the `Icon` property probably messed something up. There's a working example of this in ActionButton though that one operates on external Controls which is probably the reason it works. I think I'm fine with this then, at the very least internal `set_icon` method is not run if given icon is the same as previous so no redraw call every frame.

**Comment by Kasterisk** [comment]
Still a couple of pixels close, here's the built-in checkbutton for comparison:
checkbutton

Actually you can tweak the Button's icon and text separation from thrive theme file in `Buttons/constants/hseparation` instead of manually altering the textures size to get an offset.

File: src/gui_common/CustomCheckBox.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
+﻿using Godot;
+
+/// <summary>
+///   A customized check box that changes icon when hovered / clicked.
+/// </summary>
+public class CustomCheckBox : ToolButton
```
Comment (Kasterisk) [inline src/gui_common/CustomCheckBox.cs:6]: Shouldn't `Button` as a base class is sufficient?

File: src/gui_common/CustomCheckBox.cs [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    private bool pressing;
+
+    public override void _Ready()
+    {
+        unpressedNormal = GD.Load<Texture>("res://assets/textures/gui/bevel/checkA.png");
+        unpressedHover = GD.Load<Texture>("res://assets/textures/gui/bevel/checkAhover.png");
+        unpressedClick = GD.Load<Texture>("res://assets/textures/gui/bevel/checkAclick.png");
+        pressedNormal = GD.Load<Texture>("res://assets/textures/gui/bevel/checkB.png");
+        pressedHover = GD.Load<Texture>("res://assets/textures/gui/bevel/checkBhover.png");
+        pressedClick = GD.Load<Texture>("res://assets/textures/gui/bevel/checkBclick.png");
```
Comment (Kasterisk) [inline src/gui_common/CustomCheckBox.cs:24]: There's a better way to do this. Add a custom item to thrive theme file for CheckButton and use `GetIcon` to get that icon when updating this button's display icon.

File: src/gui_common/CustomCheckBox.cs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        // we have to capture mouse event.
+        if (@event is InputEventMouseButton { ButtonIndex: (int)ButtonList.Left } mouseEvent)
+        {
+            pressing = mouseEvent.Pressed;
+        }
+
+        base._GuiInput(@event);
+    }
+
+    private void UpdateIcon()
```
Comment (Kasterisk) [inline src/gui_common/CustomCheckBox.cs:48]: The logic in this method can be organized better with `GetDrawMode()`. See the [DrawMode](https://docs.godotengine.org/en/stable/classes/class_basebutton.html#enumerations) enum.

**Review by Kasterisk — CHANGES_REQUESTED** [review]

File: src/gui_common/CustomCheckBox.cs [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+            return;
+
+        UpdateIcon();
+        base._Process(delta);
+    }
+
+    public override void _GuiInput(InputEvent @event)
+    {
+        // Only when button's press state changes does Godot call _Pressed(), so to show a different icon when clicked,
+        // we have to capture mouse event.
```
Comment (Kasterisk) [inline src/gui_common/CustomCheckBox.cs:39]: Not a change request but want to mention that when _Pressed() is called depends on the button's `ActionMode` which determines when the button is considered click. There are two: button release and button press, so on button press mode, _Pressed() is called right on mouse click and not release.

**Comment by 84634E1A607A** [comment]
> from thrive theme file

Actually I found another in Button -> Custom Constants to avoid unexpected changes.

File: src/gui_common/CustomCheckBox.cs [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+        unpressedNormal = GetIcon("UnpressedNormal", "CheckBox");
+        unpressedHovered = GetIcon("UnpressedHovered", "CheckBox");
+        unpressedClicked = GetIcon("UnpressedClicked", "CheckBox");
+        pressedNormal = GetIcon("PressedNormal", "CheckBox");
+        pressedHovered = GetIcon("PressedHovered", "CheckBox");
+        pressedClicked = GetIcon("PressedClicked", "CheckBox");
+        UpdateIcon();
+    }
+
+    public override void _Process(float delta)
```
Comment (Kasterisk) [inline src/gui_common/CustomCheckBox.cs:42]: Turns out there's actually a way to properly update the icons in `_Draw` method. You just need to defer the custom icon update call so it doesn't conflict with the internal one. Can be done by using `Invoke.Queue` method.

File: src/gui_common/CustomCheckBox.cs [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    public override void _Ready()
+    {
+        unpressedNormal = GetIcon("UnpressedNormal", "CheckBox");
+        unpressedHovered = GetIcon("UnpressedHovered", "CheckBox");
+        unpressedClicked = GetIcon("UnpressedClicked", "CheckBox");
+        pressedNormal = GetIcon("PressedNormal", "CheckBox");
+        pressedHovered = GetIcon("PressedHovered", "CheckBox");
+        pressedClicked = GetIcon("PressedClicked", "CheckBox");
```
Comment (Kasterisk) [inline src/gui_common/CustomCheckBox.cs:38]: There is the default `checked` and `unchecked` icon, perhaps those can be reused for `pressedNormal` and `unpressedNormal`.

**Review by Kasterisk — CHANGES_REQUESTED** [review]
Noticed there's a considerable spacing between the custom checkboxes. See comparison:

Probably needs some fiddling with the internal margins to get them look closer to the default checkboxes.

File: src/gui_common/CustomCheckBox.cs (hunk 16 above)
Comment (hhyyrylainen) [reply src/gui_common/CustomCheckBox.cs:42]: What was the problem regarding that when _Draw is just overridden and looks like this:
```
protected override void _Draw(){
base._Draw();
// do our custom stuff here
```
to override what happens by default?

File: src/gui_common/CustomCheckBox.cs (hunk 17 above)
Comment (84634E1A607A) [reply src/gui_common/CustomCheckBox.cs:38]: I prefer the current one that all custom icons are named consistently.

**Comment by 84634E1A607A** [comment]
> Probably needs some fiddling

I've failed to change margins. I hide the button background, and the space between them is partly background. They are decided by Godot and I'm not sure how to override it.
image

File: src/gui_common/CustomCheckBox.cs (hunk 16 above)
Comment (Kasterisk) [reply src/gui_common/CustomCheckBox.cs]: There's no problem with the `_Draw` override itself, the problem is when setting `Icon` property in the `_Draw` method. So what happens is:
1. mouse hovers button area,
2. button redraws (which includes text and icon) and `draw` notification is propagated,
4. we set `Icon` with custom hovered icon in `_Draw`
5. `Icon` is set thus `draw` is propagated once more

All this happens concurrently and the two icon redraw call conflicts with each other so by deferring the `UpdateIcon` call we kind of separates the initial hovered redraw and the subsequent custom icon redraw.

**Review by Kasterisk — CHANGES_REQUESTED** [review]
Override the custom checkbox's styleboxes with StyleBoxEmpty
sb
Apparently the margins comes from the styleboxes.

File: src/gui_common/CustomCheckBox.cs [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        Icon = currentCheckState switch
+        {
+            CheckState.Disabled => Pressed ? pressedNormal : unpressedNormal,
+            CheckState.UnpressedNormal => unpressedNormal,
+            CheckState.UnpressedHovered => unpressedHovered,
+            CheckState.UnpressedClicked => unpressedClicked,
+            CheckState.PressedNormal => pressedNormal,
+            CheckState.PressedHovered => pressedHovered,
+            CheckState.PressedClicked => pressedClicked,
+            _ => throw new ArgumentOutOfRangeException(),
```
Comment (Kasterisk) [inline src/gui_common/CustomCheckBox.cs:89]: Add an exception message.

**Review by Kasterisk — CHANGES_REQUESTED** [review]

**Comment by 84634E1A607A** [comment]
Well, after renaming my local branch from tweaked_check_box to custom_check_box, this pr is closed automatically...

**Comment by hhyyrylainen** [comment]
Well you deleted the remote branch this PR was created from.

**Pull request closed without merging**